utils/s2n_random.c

```diff
@@ -346,9 +373,14 @@ int s2n_openssl_compat_rand(unsigned char *buf, int num)
     struct s2n_blob out = { 0 };
     POSIX_GUARD(s2n_blob_init(&out, buf, num));
 
+    if (s2n_is_in_fips_mode()) {
+        return 0;
+    }
```

Comment:
I'm still not sure about this.

S2N_LIBCRYPTO_SUPPORTS_CUSTOM_RAND and s2n_is_in_fips_mode() are technically not checking the same thing. They could theoretically disagree. In particular, if correctly implemented, S2N_LIBCRYPTO_SUPPORTS_CUSTOM_RAND should really only be undefined/false if the methods required to set up custom random don't exist, not just if we don't want to use them. I'm not sure whether that's true, because we don't properly use feature detection to set it: https://github.com/aws/s2n-tls/blob/5c22c5d5abaca4736bc97461e0cdd365e1b33730/crypto/s2n_openssl.h#L53C1-L57 But if we did shift to feature detection, this check might break. For example, I think the APIs technically exist in awslc-fips, they're just no-ops.

The right answer might be to modify [s2n_rand_init](https://github.com/aws/s2n-tls/blob/5c22c5d5abaca4736bc97461e0cdd365e1b33730/utils/s2n_random.c#L393-L399) to not attempt to set the callback if fips is enabled, regardless of whether the libcrypto supports custom rand.

Reply:
Yeah, that makes sense. The check in s2n_rand_init is what I had originally. I reverted https://github.com/aws/s2n-tls/commit/eeeb4c37a13e22ce5e27a54785e96efd63683044 to move the check back to there.

Reply:
I also removed the test that makes sure s2n-tls isn't in FIPS mode if S2N_LIBCRYPTO_SUPPORTS_CUSTOM_RAND is true, in case we switch this to feature detection. This test isn't needed anymore since the libcrypto implementation won't be used in FIPS mode, even if S2N_LIBCRYPTO_SUPPORTS_CUSTOM_RAND is true.